### DISCUSSION TITLE
Welcome to Third Room!

### DISCUSSION THREAD

**Discussion by robertlong** [matrix-org/thirdroom#20] [Announcements]
![image](https://user-images.githubusercontent.com/1753624/153054119-eeda48d6-5ab8-4fb7-aecb-92e4f2867c6d.png)

**Note: The contents of this document are subject to change as we further plan out the Third Room roadmap**

Third Room is a decentralized metaverse platform built on top of the [Matrix protocol](https://matrix.org). The project is governed by the Matrix.org Foundation. It is entirely open-source and developed in the open. If you want to get involved in development, you can join us in the [#thirdroom-dev:matrix.org](https://matrix.to/#/%23thirdroom-dev:matrix.org) Matrix room.

Third Room enables people to discover and create virtual worlds while protecting their privacy and giving them ownership over their data. We use open standards to ensure that both the 3d content and social networking data is portable and interoperable with other platforms. We empower creators to design and program rich and interactive virtual worlds and avatars using a wide variety of 3D content creation tools and harnessing the power of the glTF file format.

The Third Room client developed in this repository is only one implementation of a possible Third Room client and we welcome additional web or native clients to implement the protocols we design together. Third Room itself is designed around the Matrix protocol, but pieces of this project can be utilized in other 3D web applications including other metaverse platforms. We intend to build the client in a modular fashion, improving the overall landscape for the 3D web ecosystem and being a good participant in the open source community that we are benefitting from.

We will be a good participant in the Open Metaverse. We believe that there is one metaverse just as there is one internet. Third Room is just one piece of the overall metaverse. We will embrace and encourage interoperability instead of accepting fragmentation.

Third Room at its core is focused on creating spaces for people and communities to feel safe. We will foster an open community that is kind, respectful, and accepting of others. We will utilize Matrix's existing moderation tools and extend them to best protect our community members against trolls and harassment. We will put users in control of their own privacy and always design with a privacy and safety first mindset.

## Problems

- Virtual worlds are disconnected which makes multiplayer gaming and switching activities or applications difficult and clunky.
- Some virtual world platforms are monetized with an ad based model that uses sensitive user data to recommend ads.
- Other virtual world platforms are monetized via NFTs which can be inaccessible due to cryptocurrency transaction fees, some are built on proof of work chains that require vast amounts of electricity, and all operate on the concept of artificial scarcity which has been co-opted by many groups in various speculative schemes.
- There is high demand from audiences for free user-generated content games on platforms where creators make next to nothing. The audiences and creators for these platforms heavily skew younger and are being exploited by platform operators.
- The quality of these user-generated games can be high but most are not due to the platform restrictions especially around monetization opportunities. Creators cannot afford to create indie/AAA content.

## The Pitch

- Third Room will be built on top of the Matrix protocol allowing for decentralized hosting of virtual worlds.
- Worlds can be traversed quickly by changing rooms.
- Your identity, network, and other shared data will be brought with you between these worlds.
- Worlds will have spatial voice chat and game networking that can be made end-to-end encrypted so that you know you are only sending data to the people in the room.
- Creators of all skill levels will be able to create worlds using either a beginner-friendly in-world collaborative tool or industry-standard 3D tools.
- Creators will be able to use the programming language of their choice compiled to WebAssembly to add interactivity to their worlds.
- Creators will also be able to monetize their worlds through marketplaces for 3D content including avatars, environments, and props as well as scripts. They also can monetize room or server membership.

## Design Concepts

Designs subject to change. Environment by [James Lago](https://www.artstation.com/artwork/PoKrly). Avatars by [Ready Player Me](https://readyplayer.me/).

![dash](https://user-images.githubusercontent.com/1753624/153054359-56579985-a6b3-41f0-8cd1-a8ed1cbe55e6.png)

![loading](https://user-images.githubusercontent.com/1753624/153054403-556f1403-93ff-48b8-99f3-7517c1be32cf.png)

![in room](https://user-images.githubusercontent.com/1753624/153054462-24b31800-a086-4c34-b7bc-8b279af8faa7.png)

![in room-1](https://user-images.githubusercontent.com/1753624/153054498-b52f520d-b46b-4a31-8ca6-98096eb876b7.png)

![in room-2](https://user-images.githubusercontent.com/1753624/153054525-c92d2b89-e662-47ea-8536-ae6ebe006302.png)

## Media

[OMI Demo Night Talk](https://www.youtube.com/watch?v=e26UJRCGfGk&t=2266s)
[FOSSDEM 2022 Native Group VoIP and Metaverse on Matrix](https://youtu.be/eLQauo7IMnM?t=412)

## First Steps / Minimum Viable Product

- We will be developing a more polished and feature-complete version of the original Third Room demo.
- Users will be able to sign in to their Matrix account using OpenID Connect
- They will be able to create Matrix rooms containing 3D virtual worlds by choosing from one of the provided scenes or uploading their own glTF scene. The room's scene will be stored in Matrix. A screenshot/thumbnail of this scene will also be stored in Matrix to show in room listings and when loading the scene. This scene may include a Web Assembly script which will have some scripting capabilities to add interactivity to the room.
- Users will be able to view a list of their joined rooms, search for public rooms, and invite others to their rooms.
- They will be able to customize and pick their avatar from an existing set or upload their own using either the Mozilla Hubs avatar skeleton or VRM skeleton, both represented as a glTF avatar. Their profile containing this avatar will be stored in Matrix.
- Users will be able to use text and voice chat to talk to other users in the room. This will be based on top of Matrix and the Matrix Group Calls work. Connections to other users in the room will be made via P2P WebRTC connections and SFUs will not be in place for the MVP.
- Users will be able to walk around these virtual worlds, embodying their avatar in both 3rd and 1st person. Avatars with and without legs will be supported. XR support will not be in the MVP.
- Users will be able to see the list of users in the room and block, kick, or ban other users depending on their permissions in the room.
- Users will be able to quickly jump between scenes via links inside rooms or by clicking on rooms in the room list.
- Users will be able to access rooms, avatars, and various assets across multiple thirdroom servers. Federation will be enabled from the start.

## Core Capabilities for MVP

- Matrix authentication via OpenID Connect
- Matrix account registration
- Avatar upload, thumbnail creation, and representation in Matrix
- Matrix room listings
- Matrix room creation
- Scene upload, thumbnail creation, and representation in Matrix
- Multi-party spatialized voice chat
- Matrix text chat
- In-room participant list
- In-room moderation tools (block/ignore, kick, ban, mute)
- WebRTC datachannel-based game networking
- Basic avatar customization (TBD)
- Hubs avatar skeleton support
- VRM avatar skeleton support
- Avatar animation system
- Physics system
- Physics-based character controller
- Input system (keyboard/mouse only)
- Scene transition system (following links / room listing)
- Matrix Federation support
- glTF scene loader supporting OMI and Thirdroom-specific extensions
- MVP of WASM scripting support
- Scene/avatar authoring pipelines via Blender
- 1st party and community content for scenes/avatars
- Alpha plugin support

## Things we are explicitly avoiding for MVP

- XR support
- Inventory/trading systems
- Friends list
- SFU / High capacity rooms
- Mobile, Touchscreen, and Gamepad support
- Desktop app
- Text/Video/Voice only rooms
- Content marketplaces
- Emotes
- Bot API / authoritative server support
- Bridging with other platforms

## Tech Stack

- Synapse: Matrix Homeserver
- Hydrogen SDK: Lightweight, performance-focused, Matrix Client SDK
- React: Frontend UI framework
- Vite: fast/low-configuration JS bundler/dev server
- Three.js: WebGL 3D rendering library
- Rapier.js: Rust/WASM physics library with first-class, well-documented JS API
- bitECS: Fast, data-oriented, JS ECS library
- three-omi: Support for loading OMI glTF extensions
- [three-vrm](https://github.com/pixiv/three-vrm): Support for loading VRM avatars

TBD:

- CSS Framework / Component Library
  - CSS Modules
    - PostCSS
    - SASS
  - Tailwind CSS
  - Radix UI
  - CSS-in-JS
    - styled-components
    - stitches
- [Postprocessing](https://github.com/vanruesc/postprocessing) Three.js postprocessing effects
- [three-mesh-bvh](https://github.com/gkjohnson/three-mesh-bvh) Three.js mesh raycasting

To Investigate:

- [Ossos (Character Animation)](https://github.com/sketchpunklabs/ossos)

# Key Stakeholders

- Matrix.org
  - Robert Long: Project Lead / Frontend / 3D / VoIP Engineer
  - Nate Martin: 3D Engineer / bitECS Creator / WebECS Project Lead
  - Bruno Windels: Frontend / Matrix Client Engineer / Hydrogen Lead
  - Matthew Hodgson: Element CEO/CTO / Matrix.org Lead & Guardian
- Cinny.in
  - Ajay Bura: Frontend Engineer / Cinny Lead
- [XP Portal](https://xpportal.io/)
  - Anthony Burchell: Frontend / 3D Engineer
  - Ben Moore: Frontend / 3D Engineer

# MVP Content

What environment / experience can we build that gives users something that is:

- Useful and/or entertaining
- Genuinely a good experience in 3D
- Fits the vibe we would like to set for Third Room as a product
  - Fun
  - Relaxing
  - Social
  - Creative
- Within a reasonable scope for our MVP
- Works within the boundaries we'd like to set for the WASM API

How do we want to go about getting content for this environment?
- 3D artists to contract with?
- Technical + visual artist to hire?

**Comment by TheFrenchGhosty** [comment]
Please don't, just, NO.

I really thought Matrix was a good thing, I guess I'll have to move to XMPP.


https://xpportal.io/pricing-page/ thanks for the laugh (it's PER MONTH).

**Reply by tezlm** [reply]
I really don't like the idea of a metaverse, but when everyone else inevitably gets on the hype train and drags us along... well, it's better to do so on an open protocol rather than a closed one.

**Reply by hrj** [reply]
Speaking of laughs, never was a music video more apt:
https://www.youtube.com/watch?v=QcD_YXCxxZM

**Reply by TheFrenchGhosty** [reply]
@tezlm 

> I really don't like the idea of a metaverse, but when everyone else inevitably gets on the hype train and drags us along... well, it's better to do so on an open protocol rather than a closed one.

It's better to not do it, at all.

**Reply by ara4n** [reply]
Just because the word 'metaverse' has been tainted by Facebook and others doesn't mean that the idea of a decentralised spatial environment built on open standards is a bad idea.  After all, the metaverse is basically here already in the form of Google Docs, Figma, HackMD and protocols like Matrix and ActivityPub - we're just lacking an open interoperable means to link it together, and that's what we're trying to fix.  For me, it's a logical extension of Matrix as the open real-time comms layer of the open web to support richer kinds of collaboration than just chat & voice/video.

An explicit goal of Third Room is to provide a basis for building virtual worlds in a way which isn't locked in proprietary silos like FB, or contingent on cryptocurrencies - and is open and available to everyone just like the open Web is.

Obviously, if you don't like this idea, you can simply ignore it and not use it.  Or, even better, you can move to XMPP :)

**Reply by MxSelfDestruct** [reply]
as long as this has nothing to do with cryptocurrency/NFTs this seems pretty cool tbh. the use of ready player me in promo material though does concern me.

**Reply by robertlong** [reply]
Personally, I don't really like the term metaverse. It has a bunch of baggage that comes with it. I like the Spatial Web, or the Wider Web as Trevor Flowers has coined. Both of those terms really center on the web as the only platform, which I don't think is the end game for Third Room. We use the term metaverse because it's the most widely understood term for describing interoperable virtual worlds and the addition of a social layer onto the internet. In a lot of ways, Third Room is about adding the ability to host 3D worlds on Matrix and potentially other applications and usecases that center around collaboration and realtime social applications.

@MxSelfDestruct yeah it concerns me too. To the point that I don't think we'll consider them as a partner. Instead we'd like to work with the community and potentially commission avatars from independent artists. We want to see people paid for their work without having to resort to selling them as NFTs. If you know of anyone who would be interested, please send them our way!

**Reply by robertlong** [reply]
@hrj I love Khruangbin, the song/video are definitely an inspiration for this 🙂 as well as the Third Place
https://en.wikipedia.org/wiki/Third_place

**Reply by lalalune** [reply]
There is a real need for lightweight, decentralized 3D spatial web platform and this fills the need better than anything else right now. It's probably the only decentralized spatial web project out there that I have seen that isn't shoving blockchain in your face. I strongly believe this project is in everyone's best interest.

**Reply by ghost** [reply]
I agree that language can be abused in an attempt to gain control and power.  My hope is that a clear, clean architecture would be the basis for decentralization of many things.  Communication and connectivity are key.  The choice of MVP need not bias opinion on the underlying goal, IMHO.

**Reply by jreoka** [reply]
This is a cool idea to just use to hang out and chill with friends. No matter what you do there will always be people with a negative opinion. You just gotta ignore it and keep moving forward.

**Comment by 4www** [comment]
There is also the https://aframe.io library, for a declarative approach, where every message (that includes HTML), can become a word event (or a word itself), from a chat input (textarea) field.

Maybe a combination of user defined third spaces on the decentralized Matrix network can be a nice counter-offer to the existing centralized capitalist dominance.

Right now if the alternative multiverse of events through computer system and their state is already real, what we lack is visualizations of these.

The URI seems central, as it allow to share a query of the matrix state and from there decide what to be rendered.

This connects well through the URI scheme for Matrix:
- https://github.com/matrix-org/matrix-spec-proposals/blob/old_master/proposals/2312-matrix-uri.md#uri-scheme-for-matrix
- https://github.com/matrix-org/matrix-spec-proposals/issues/455

If this idea of "alternative spacial verses" is supported and part of the Matrix.org idea, it could be even more connected to treating each rooms, servers, events, as "unix" files.

Notes:
- matrix client "that looks like a file explorer", but can render rooms/events of different types, including 3d/vr.
- https://www.emacswiki.org/emacs/TrampMode (maybe through https://github.com/alphapapa/ement.el)
- https://en.wikipedia.org/wiki/Plan_9_from_Bell_Labs

Last, there is a concern about the amount of data that is being sent into matrix (not that other systems have a better state), but since every event stored in matrix is associated to a user id, it can give a lot of information in case of an accidental leak of data.

**Reply by DavinciDreams** [reply]
I have been thinking about this approach, and I have seen and alpha of blogs that you can hop into like a clubhouse room and I am interested in being able to host rooms that are persistent on a given topic, would love to chat further if you are amenable! DavinciDreams is my github.

**Comment by Extarys** [comment]
Not sure if I love or hate the idea, but I believe there is much more to do in matrix/element before trying to implement this. I know voice calls and voice rooms are comming. Would be nice to have a tool for people managing servers, clearing out old messages, number of active users, avg disk used per user or similar. Even a small game API would be great. Auto deleting messages like signals, attachments that can be only seen X number of times and then deleted from the server, more granular roles/permission.

I mean, I can see myself talking to some friends near a nice water pond in a forest, depends on how far that project goes.

I also 100% agree with @user-09876 :+1: 
> Last, there is a concern about the amount of data that is being sent into matrix (not that other systems have a better state), but since every event stored in matrix is associated to a user id, it can give a lot of information in case of an accidental leak of data.

Matrix/Element need more tools to help both server admins and the end user to secure their data. I read so many posts/blog/opinion piece about why Matrix protocol/Element cannot be trusted and I believe some more features can be implemented to, not win but at least ease some users.

**Reply by ghost** [reply]
For the uninitiated, could you perhaps post a summary of points (links) regarding lack of trust?

**Reply by Avi0n** [reply]
>Would be nice to have a tool for people managing servers, clearing out old messages, number of active users, avg disk used per user or similar.

The [Synapse-admin](https://github.com/Awesome-Technologies/synapse-admin) web UI covers some of these concerns.

**Reply by Extarys** [reply]
@mintyDave7s
- https://gitlab.com/libremonde-org/papers/research/privacy-matrix.org
- I can't seem to find the other post/uri, I'll update if I find it later tonight
- 
Now to be fair, here is a commented version of the same document above by Matrix: https://matrix.org/~matthew/Response_to_-_Notes_on_privacy_and_data_collection_of_Matrix.pdf

I didn't post this to stir anything, just that there is much to do with Matrix and the VR, Metaverse NFT overdrive is a niche only the riches and the dumbs can afford (NFT scams all around, Meta (the company), and I'm way too poor to buy a VR headset. A nice 3D world to hang it is dope but already exists - a matrix way would indeed be cool if the development/expansion by 3rd party users is possible with open source software. But at this point the project is too early for anything set in stone.

@avion540 :O Thank you! Hope it works with dendrite, I'll look into it tomorrow :D

**Comment by superboum** [comment]
When I have first seen your project name, "Third Room", I thought to Ray Oldenburg's [Third Place](https://en.wikipedia.org/wiki/Third_place). In the comment, I have seen @robertlong making the connection, so I think this is an intended reference.

But at the same time, your introductory speech is centered around metaverse and technology. I know that you have economic incentives, but at the same time, I think you are never better that when you have a critical speech. For example, I think to your discussion with [LQDN](https://www.laquadrature.net/en/) about moderation that was very refreshing. 

I would really like to read a post from you introducing Ray Oldenburg's work, explaining critically why such project matters, what is the essence of your project, if possible you would invite some academics/experts. In the end, it could even help to reframe/reshape the "metaverse" buzzword to something more desirable, which would be another huge achievement for Matrix :-)

**Reply by erlend-sh** [reply]
Here’s a great writeup on The Third Place:
https://invisibleup.com/articles/31/

Terence McKenna is also an excellent source of inspiration on this topic:
https://youtu.be/dyUopIVjvHg

**Comment by erlend-sh** [comment]
Love this initiative! 🙌 

I would however appreciate a cleaner separation between the (1) Matrix capabilities and (2) game engine capabilities needed to support the elusive *metaverse*, since 3D is by no means an essential part of the metaverse. Some good 2D examples include:
- https://www.gather.town/
- https://github.com/thecodingmachine/workadventure
  (hello @moufmouf & @gparant)

**Comment by jreoka** [comment]
This looks so cool! Can't wait! <3

**Comment by Roki100** [comment]
if only element had such clean design...

**Comment by yunkuangao** [comment]
Cool idea!

**Comment by 32bitx64bit** [comment]
Just no, anything metaverse is stupid. 

I really hope this DOESNT become a thing.
